# G-GArden

G-GArden is a graphical user interface (GUI) based tool for primer design in Golden Gate Assembly (GGA) strategies. It was primarily developed to construct plasmids for the expression of synthetic small RNAs (sRNAs) in bacteria.

## Background
Using GGA, multiple double-stranded DNA fragments, each containing specific unique overhangs, are combined in a one-pot, one-step reaction to produce plasmids with one or more transcriptional units (TUs). These TUs may encode sRNAs, which can be used to repress any target gene of interest in a host bacterium. Typical sRNA TUs consist of (i) a promoter, (ii) a sequence for an sRNA seed region that enables binding to the target mRNA, and (iii) a sequence for an sRNA scaffold.

Precise planning is needed to ensure that all fragment overhangs are unique and matching, which will enable ligation (assembly) of fragments in the correct order. The unique overhangs can be created in several ways. For smaller fragments, oligodeoxynucleotides can be used for annealing reactions. In this case, forward and reverse oligodeoxynucleotides contain matching sequences except for the unique overhangs at their 5’ ends. For larger fragments, PCR primers can be used to amplify the region of interest. In order to release the correct fragment from the amplified region, PCR primers contain recognition sites for type IIS restriction enzymes at their 5’ ends. Type IIS enzymes cleave DNA at a defined distance outside of their recognition site. Primers are designed in a way that recognition sites are removed by the cleavage. Type IIS restriction enzymes are also used to linearize acceptor plasmids. Because many fragments might be combined, the complexity of planning skyrockets with raising number of fragments.

The G-GArden tool was developed to simplify planning of GGA strategies. The tool uses a collection  of customized plasmids and sequences, from which the user chooses the fragments for assembly. The tool will check for recognition sites of type IIS restriction enzymes and suggest an assembly strategy based on unique overhangs. Specific problems of the assembly strategy will be reported. If the assembly strategy is feasible, the user will receive several output files, including visualization of the strategy, oligodeoxynucleotides/ primers for ordering and plasmid maps (*.gbk* files). The GUI provides selection of fragments for single and multiple TU constructs, multi-fragment selection for parallel and combinatorial design, input options for addition of new sequences to the database, options for using *primer3*, and compatibility with the SEEDling tool to feed in optimal seed-scaffold combinations for the target gene of interest.


## Installation/Run
The G-GArden tool was created with [*PyInstaller*](https://pyinstaller.org/en/stable/). Hence, users do not need to preinstall anything. Simply download the Zip folder (versions for both Linux and Windows are available). To start your journey, navigate through *dist* in *ggarden* and start the *ggarden* executable by double-clicking. At this point, the GUI should open and you are ready to go.


## Usage
The G-GArden GUI contains a multitude of different widgets. The **Add**-Button in the upper left is the most important part of the GUI. The widget enables selection of fragments for GGA. Fragments are retrieved from the *collection* file (*saved_data.csv*). The first fragment is always an acceptor plasmid that is selected from a dropdown. Acceptor plasmids are Golden Gate-compatible plasmids that contain recognition sites for type IIS restriction enzymes. Once the acceptor plasmid is selected (multi-selection is also available), more fragments can be added and selected from additional dropdowns. For these fragments, a second widget is available, which allows you to specify the fragment creation mode (**PCR** or **Anneal**). The default option **Auto** will determine the fragment creation mode based on the **Maximum Annealing Length** parameter. Once you have selected fragments for assembly, a third widget will appear (**overhang** widget). The third widget is a text input field, in which you can specify the sense-strand overhang that is used for ligation of two fragments. This widget will only appear if the corresponding overhang can be customized. Please be aware that the tool will try to avoid creation of scars in the final constructs, but specification of custom overhangs will create scars. To remove a previously added dropdown, the **Del**-Button can be used. You can also access the sequence of all selected fragments by using the **Get Seq**-Button. A new window will open, in which you can copy individual sequences.

The main window of the G-GArden GUI contains several additional widgets for changing parameters:
-	**Restriction Enzyme** (enzyme name, dropdown) - The type IIS restriction enzyme used for cutting of the acceptor plasmid and PCR products.
-	**Overhang Length** (none) - The length of the overhangs created by the type IIS restriction enzyme. The overhang length is only displayed and cannot be changed, because it is based on the selected restriction enzyme.
-	**Maximum Annealing Length** (number) - The maximum length of a fragment that is allowed to be created by annealing of oligodeoxynucleotides, in case Auto is selected as fragment creation mode. The maximum length includes overhangs.
-	**PCR Primer Length** (number) - Length of the PCR primers for amplification of larger fragments. If *primer3* is applied, this value can deviate by 5 bases to enable optimal primer design.
-	**Melting Temperature** (number, °C) - Only applicable if *primer3* is used. The melting temperature of the PCR primers suggested by *primer3*. This value can deviate by 5°C.
-	**Bases for Gap** (A, C ,G, T) - Bases that are added at the start of any PCR primer in front of the recognition site for the type IIS restriction enzyme. A subset of these bases will be used to fill the gap between the recognition site and the cut-site.
-	**Project Name** (file name, without ending) - Shared name of output files and directory of the run.

After selection of all desired fragments and specification of parameter settings, you may want to check the overhangs using the **Overhangs**-Button. A new window will open, in which the overall fidelity of the assembly reaction is given. The fidelity depends on the assembly parameters, which can be selected from a dropdown at the top (T4 ligase reaction for 1 or 18 hours at 25 or 37°C). Furthermore, a color code indicates the quality of individual overhangs. Green indicates that the overhang can be used without further refinement (by default, plasmid overhangs are green). Yellow indicates low-fidelity overhangs, which can be optimized if desired. Red indicates duplicate overhangs, an issue that has to be fixed. The fidelity computation is based on experimental fidelity estimation (Potapov et al., 2018a and Potapov et al., 2018b). For optimization of overhangs, three modes are available from the slider at the top: (1) left position (grey) aims at optimizing all overhangs for the highest assembly fidelity possible, (2) middle position (red) will only replace duplicate overhangs, (3) right position (red/yellow) will replace red and yellow overhangs. After selection of the mode, all mandatory overhangs are stored in your clipboard by left-clicking the corresponding text line (a successful copy is indicated by random color change). The **Optimize**-Button will direct you to the NEBridge GetSet™ Tool (https://ggtools.neb.com/getset/run.cgi), where you can select the assembly parameters (“Ligation conditions”) and the total “Number of overhangs”. The mandatory overhangs are now pasted from the clipboard into the “Required overhangs” field. After submitting the job, the NEBridge GetSet™ Tool will suggest all missing (optimized) overhangs. The missing overhangs can be copy-pasted in the corresponding overhang widgets in the main window of the G-GArden GUI. Please be aware that the default setting of G-GArden avoids scars in the final constructs. By choosing custom overhangs (e.g., by choosing optimized overhangs from the NEBridge GetSet™ Tool), these overhangs will be present as scars in the final constructs.

After completion of overhang selection and optimization, the **Run**-Button in the bottom right is used to generate the output. If the cloning strategy is feasible, the **Output Directory**-Button will appear in the middle right. Otherwise, an error message will appear in the **Exception Display** area in the middle left. The display can be reset by left-clicking the error text. The **Output Directory**-Button will open the output directory, in which folders are labeled according to the name of the cloning project (as specified in the **Project Name** widget) and according to the date/ time of the execution. The user will receive several output files: (1) plasmid maps for the final constructs (denoted xxx_gb; *.gbk* files), (2) files for visualization of overhangs for Golden Gate assembly (denoted xxx_vis; *.txt* files), (3) oligodeoxynucleotides needed for cloning of each individual construct (denoted xxx_seq; *.txt* files), (4) one table summarizing all oligodeoxynucleotides for cloning (xxx_coll; *.csv* file), and (5) one table summarizing all non-redundant oligodeoxynucleotides for cloning, which can be used for ordering (xxx_fin_coll; *.csv* file).

If you wish to add new fragments to the *saved_data.csv* file, there are three options to do so. The most straightforward option is to open the *saved_data.csv* file, in which all plasmids and fragments are saved, and edit the file via Excel or another editor of your choice. Each row has to be built in a specific order, starting with the fragment name, followed by fragment type (plasmid, promoter, seed, or scaffold), sequence of the fragment, and a path to a genbank (*.gbk*) file (genbank file is optional; only useful for acceptor plasmids). If you use an editor different than Excel, please use a semicolon as separator. Excel will take care of the separators on its own. The second option is located at the bottom left of the GUI. Note that the **Insert fragment sequence** widget will only add one fragment at a time. Simply type the fragment name, fragment type, sequence and genbank path (only for acceptor plasmids), all separated by comma, and press submit. The third option is suited for adding multiple fragments at once. To use it, click the **Load File**-Button in the lower left. In the popup window, you will see your file system. You can navigate through it to search a file containing fragments to be added. Fragments are typically added from *fasta* files. The header of each fragment has to contain the fragment name, fragment type and genbank path (only for acceptor plasmids), separated by comma. As soon as a file is selected, click the **Load**-Button to add the fragments. In addition, *csv* outputs created by the SEEDling tool can be loaded the same way.

If you are reading this, you either opened the *README.md* or clicked the **Help**-Button in the upper right of the GUI.

## Primer3
*Primer3* can be used to design your PCR primers. *Primer3* will search for optimal pairs of PCR primers using sequences upstream or downstream of the desired fragment sequence (i.e. upstream of a promoter or downstream of a scaffold). This additional sequence space has to be added to the fragment sequence. The whole, continuous sequence is then uploaded to the *saved_data.csv* file. Please note that *primer3* will only be enabled if square brackets are introduced to determine the breakpoint of fragment sequence and additional sequence (upstream or downstream sequence). Examples are given in the following picture:

![Primer3 usage](./ReadMe_graphics/primer_3_usage.png?raw=true)

[Primer3 bindings for python](https://github.com/libnano/primer3-py), [Primer3](https://github.com/primer3-org)

## License
This is a repository written under the [CC BY-NC-SA 4.0 license](https://creativecommons.org/licenses/by-nc-sa/4.0/)

## How to Cite:
DIGGER-BAC is published and can be cited as following:

Philipp N, Brinkmann C, Georg J, Schindler D, Berghoff BA (2023) DIGGER-Bac: prediction of seed regions for high-fidelity construction of synthetic small RNAs in bacteria. Bioinformatics. https://doi.org/10.1093/bioinformatics/btad285

## References
Potapov, Vladimir; Ong, Jennifer L.; Kucera, Rebecca B.; Langhorst, Bradley W.; Bilotti, Katharina; Pryor, John M. et al. (2018a): Comprehensive Profiling of Four Base Overhang Ligation Fidelity by T4 DNA Ligase and Application to DNA Assembly. In: ACS Synthetic Biology 7 (11), S. 2665–2674. DOI: 10.1021/acssynbio.8b00333.

Potapov, Vladimir; Ong, Jennifer L.; Langhorst, Bradley W.; Bilotti, Katharina; Cahoon, Dan; Canton, Barry et al. (2018b): A single-molecule sequencing assay for the comprehensive profiling of T4 DNA ligase fidelity and bias during DNA end-joining. In: Nucleic Acids Res 46 (13), e79-e79. DOI: 10.1093/nar/gky303.

